### ISSUE TITLE
Missing wslpath in PATH on syschdemd

### ISSUE THREAD

**Issue by celestefox** [nix-community/NixOS-WSL#237]
## Bug description

Having just updated NixOS-WSL for the first time in a while and built a new generation, there is no `wslpath` in my `$PATH`, running with syschdemd still instead of native systemd. I have a `/bin/wslpath` that works, but it is not in my `$PATH`. This is because #209 removed the wrapper from `systemPackages`, and I believe that this still works under native systemd, due to https://github.com/nix-community/NixOS-WSL/blob/main/modules/wsl-distro.nix#L199 inheriting all of `$PATH`, which syschdemd does not do, especially as `wsl -e /usr/bin/env` reports `PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/usr/lib/wsl/lib:` before my Windows paths, which includes `/bin`.

## To Reproduce

Build an installer from `main`, install as a new temporary distro, as the configuration of the installer does not use native systemd, see that there is no `wslpath` command if you try to run it, but `/bin/wslpath` exists and works.

**Expected behavior**

A working `wslpath` on syschdemd.

## Logs
```
wsl -d nixostmp
                      Unpacking root file system...
 366MiB 0:00:15 [23.0MiB/s] [============================================================================================================================================================================================>] 100%
Activating nix configuration...
setting up /usr/share/applications...
setting up /usr/share/icons...
setting up /etc...
setting up /bin...
Cleaning up installer files...
finding garbage collector roots...
removing stale link from '/nix/var/nix/gcroots/auto/6yfbrv8wfkyl40wq6827lamr9m29z7wg' to '/build/nix/var/nix/profiles/per-user/root/channels-1-link'
removing stale link from '/nix/var/nix/gcroots/auto/zxi8v969la5n36bg7gs016wgcif4qjhp' to '/build/nix/var/nix/profiles/system-1-link'
deleting garbage...
deleting '/nix/store/vzdmnirlszj446barmaiyv1mkla89lln-wsl-prepare'
deleting '/nix/store/qk5j2w5901d2qlp535a0z3acm6fmg3p6-tarball'
deleting '/nix/store/96vxkz49y5cnnkzfgln1nksbh0kmkwsz-installer.sh'
deleting '/nix/store/khqypqm75yd71p9xn143r528jf12r7zw-source'
deleting '/nix/store/0kmxry5la971ki1gd5zzfyqq2vq1qa7l-busybox-1.35.0'
deleting '/nix/store/2y4fz7rkdzf6rmrb1016rznwj9gg93yw-pv-1.6.20'
deleting unused links...
note: currently hard linking saves -0.00 MiB
6 store paths deleted, 367.67 MiB freed
Optimizing store...
4.63 MiB freed by hard-linking 3119 files

[nixos@nixos:/]$ wslpath
wslpath: command not found

[nixos@nixos:/]$ ls /bin
mount  sh  wslpath

[nixos@nixos:/]$ /bin/wslpath
/bin/wslpath: Invalid argument
Usage:
    -a    force result to absolute path format
    -u    translate from a Windows path to a WSL path (default)
    -w    translate from a WSL path to a Windows path
    -m    translate from a WSL path to a Windows path, with '/' instead of '\'

EX: wslpath 'c:\users'

[nixos@nixos:/]$
```

**Comment by SuperSandro2000** [comment]
Did you restart WSL after the installation? 

> wsl -e /usr/bin/env

Please run commands interactively, directly running them through WSL is usally flaky but should work better since #234.

**Comment by celestefox** [comment]
I first ran into this issue with my older install, but as I said, this happens with a fresh install of an installer from `main`. I ran that command explicitly to show the default value of `PATH` from WSL and that it contains `/bin` before NixOS/NixOS-WSL do anything, but that path entry is not kept on syschdemd on purpose, while `native-systemd` behaves differently. A `wsl --shutdown` doesn't make `wsl -d nixostmp` do anything different the second "boot", there's still no `wslpath`.

**Comment by SuperSandro2000** [comment]
I think we wanted to remove syschmd in the future anyway since native-systemd fixes some bugs, like this one.

**Comment by meskill** [comment]
It doesn't work for me with `nativeSystemd` option as well. With latest installer from main there is no `/bin` in path.

Command `wsl -d nixos -e /usr/bin/env` shows its in path though

**Comment by arichtman** [comment]
I think I have a related issue - it's completely broken the normal boot process

```
PS C:\Users\ariel> wsl
/nix/store/701bd06kvch5kqzdap5rdcwdi7pxv5mc-syschdemd/bin/.syschdemd-wrapped: line 74: getent: command not found
PS C:\Users\ariel> wsl -e /usr/bin/env
HOSTTYPE=x86_64
LANG=en_US.UTF-8
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/usr/lib/wsl/lib:/mnt/c/Program Files/PowerShell/7:/mnt/c/Windows/system32:/mnt/c/Windows:/mnt/c/Windows/System32/Wbem:/mnt/c/Windows/System32/WindowsPowerShell/v1.0/:/mnt/c/Windows/System32/OpenSSH/:/mnt/c/Program Files (x86)/dotnet/:/mnt/c/Program Files/dotnet/:/mnt/c/Windows/system32/config/systemprofile/AppData/Local/Microsoft/WindowsApps:/mnt/c/Program Files (x86)/NVIDIA Corporation/PhysX/Common:/mnt/c/Program Files/usbipd-win/:/mnt/c/Program Files/PowerShell/7/:/mnt/c/Users/ariel/scoop/apps/vscode/current/bin:/mnt/c/Users/ariel/scoop/apps/gpg/current/bin:/mnt/c/Users/ariel/scoop/apps/openssl/current/bin:/mnt/c/Users/ariel/scoop/apps/corretto-jdk/current/bin:/mnt/c/Users/ariel/scoop/apps/python/current:/mnt/c/Users/ariel/scoop/shims:/mnt/c/Users/ariel/AppData/Local/Microsoft/WindowsApps:/mnt/c/Program Files/AESCrypt:/mnt/c/Users/ariel/AppData/Local/Keybase/
TERM=xterm-256color
XDG_RUNTIME_DIR=/mnt/wslg/runtime-dir
DISPLAY=:0
WAYLAND_DISPLAY=wayland-0
PULSE_SERVER=unix:/mnt/wslg/PulseServer
WSL2_GUI_APPS_ENABLED=1
WSLENV=
WSL_INTEROP=/run/WSL/101987_interop
NAME=nixos
HOME=/root
USER=root
LOGNAME=root
SHELL=/usr/bin/env
WSL_DISTRO_NAME=nix
PS C:\Users\ariel>
```

**Comment by joernweissenborn** [comment]
I got the same issue . The first startup after import is fine, any subsequent boot results in the same error:

```
PS D:\> wsl --import NixOS .\NixOS\ nixos-wsl-installer.tar.gz --version 2
Import in progress, this may take a few minutes.
The operation completed successfully.
PS D:\> wsl -d NixOS
Unpacking root file system...
1.19GiB 0:00:48 [25.2MiB/s] [===========================================================================================================================================================================================================================================================================>] 100%
Activating nix configuration...
setting up /usr/share/applications...
setting up /usr/share/icons...
setting up /etc...
setting up /bin...
Cleaning up installer files...
finding garbage collector roots...
removing stale link from '/nix/var/nix/gcroots/auto/zxi8v969la5n36bg7gs016wgcif4qjhp' to '/build/nix/var/nix/profiles/system-1-link'
removing stale link from '/nix/var/nix/gcroots/auto/6yfbrv8wfkyl40wq6827lamr9m29z7wg' to '/build/nix/var/nix/profiles/per-user/root/channels-1-link'
deleting garbage...
deleting '/nix/store/bwlcjak9cgn2f359g6w9vxc64ys0qypc-tarball'
deleting '/nix/store/hvlyywx5lqnk3bjj8d8plc4y5cwg2xj6-pv-1.6.20'
deleting '/nix/store/n9znd98f2jy1x4qpliy5znlnz29xk2pq-installer.sh'
deleting '/nix/store/q4rjlwqgivvj9k5awpkk6bwpqbqx3k9n-wsl-prepare'
deleting '/nix/store/fzigg1lbrsi1yf9l0lx3zvyld4p6c88w-source'
deleting unused links...
note: currently hard linking saves -0.00 MiB
5 store paths deleted, 1217.51 MiB freed
Optimizing store...
204.82 MiB freed by hard-linking 44755 files
/etc/set-environment: line 13: warning: setlocale: LC_TIME: cannot change locale (de_DE.UTF-8): No such file or directory
[zplug] Start to install 1 plugins in parallel

 ✔  Installed!            djui/alias-tips

[zplug] Elapsed time: 0.9941 sec.
 ==> Installation finished successfully!
❯ exit
PS D:\> wsl -d NixOS
/nix/store/ixw2c5hira8if9f2hvgs6pwhp086as3b-syschdemd/bin/.syschdemd-wrapped: line 74: getent: command not found
PS D:\> wsl -d NixOS -e /usr/bin/env
HOSTTYPE=x86_64
LANG=en_US.UTF-8
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/usr/lib/wsl/lib:/mnt/c/Program Files/Oculus/Support/oculus-runtime:/mnt/c/Windows/system32:/mnt/c/Windows:/mnt/c/Windows/System32/Wbem:/mnt/c/Windows/System32/WindowsPowerShell/v1.0/:/mnt/c/Windows/System32/OpenSSH/:/mnt/c/Program Files (x86)/NVIDIA Corporation/PhysX/Common:/mnt/c/Program Files/NVIDIA Corporation/NVIDIA NvDLISR:/mnt/c/Program Files/dotnet/:/mnt/c/Program Files (x86)/IVI Foundation/VISA/WinNT/Bin/:/mnt/c/Program Files/IVI Foundation/VISA/Win64/Bin/:/mnt/c/Program Files (x86)/National Instruments/Shared/LabVIEW CLI:/mnt/c/Program Files (x86)/IVI Foundation/VISA/WinNT/Bin:/mnt/c/Program Files/Git/cmd:/mnt/c/Program Files/Go/bin:/mnt/d/TDM-GCC-64/bin:/mnt/c/Users/joern/AppData/Local/Microsoft/WindowsApps:/mnt/c/Users/joern/go/bin
TERM=xterm-256color
XDG_RUNTIME_DIR=/mnt/wslg/runtime-dir
DISPLAY=:0
WAYLAND_DISPLAY=wayland-0
PULSE_SERVER=unix:/mnt/wslg/PulseServer
WSL2_GUI_APPS_ENABLED=1
WSLENV=WT_SESSION::WT_PROFILE_ID
WT_SESSION=c98f5d02-6c32-4a19-b836-b3d741464406
WT_PROFILE_ID={61c54bbd-c2c6-5271-96e7-009a87ff44bf}
WSL_INTEROP=/run/WSL/982_interop
NAME=nixos
HOME=/root
USER=root
LOGNAME=root
SHELL=/usr/bin/env
WSL_DISTRO_NAME=NixOS
```

My config from which I built the installer from is here: https://github.com/joernweissenborn/nixos-config/blob/main/hosts/tina-wsl/default.nix

I already set `stateVersion` of my nixos to 23.05, could that be an issue?

**Comment by arichtman** [comment]
I think enabling native systemd fixed it for me

https://github.com/arichtman/nix/blob/2cd654bff9d11cbdf767412845d5dc97c311e01c/modules/nixos/wsl/default.nix#L44

**Comment by joernweissenborn** [comment]
> I think enabling native systemd fixed it for me
> 
> https://github.com/arichtman/nix/blob/2cd654bff9d11cbdf767412845d5dc97c311e01c/modules/nixos/wsl/default.nix#L44

That didn't work for me, now I get immediate crash:

```
PS D:\> wsl --import NixOS .\NixOS\ nixos-wsl-installer.tar.gz --version 2
Import in progress, this may take a few minutes.
The operation completed successfully.
PS D:\> wsl -d NixOS
Catastrophic failure
Error code: Wsl/Service/CreateInstance/E_UNEXPECTED
```

I see you use docker, maybe I should try that.

**Comment by terlar** [comment]
I started seeing this issue as well after updating `nixpkgs` and running `sudo nixos-rebuild switch`:
```
• Updated input 'nix-config/nixpkgs':
    'github:NixOS/nixpkgs/3ae20aa58a6c0d1ca95c9b11f59a2d12eebc511f' (2023-06-23)
  → 'github:NixOS/nixpkgs/3c7487575d9445185249a159046cc02ff364bff8' (2023-07-06)
```

After the switch I get when trying to launch a WSL shell:
```
/nix/store/0q78jsg9h19slxaxnjhwvjm6x42fygcj-syschdemd/bin/.syschdemd-wrapped: line 74: getent: command not found

[process exited with code 127 (0x0000007f)]
You can now close this terminal with Ctrl+D, or press Enter to restart.
```

If I revert back to the previous nixpkgs version everything works fine again.

I can fix the issue by enabling native systemd, but then instead of getting logged in as my user, I get logged in as root when starting a new shell and if I `su` to my user, my user doesn't have permissions to execute anything, e.g. `ls` etc.

**Comment by arichtman** [comment]
That mirrors my experience exactly. I updated my dependencies, it broke on getent, enable native systemd and it logged me into rood. Fully terminate the VM and it works fine now.

**Comment by AtnNn** [comment]
Here's a potential workaround for the getent error:

```
PS C:\Users\atnnn> wsl
/nix/store/701bd06kvch5kqzdap5rdcwdi7pxv5mc-syschdemd/bin/.syschdemd-wrapped: line 74: getent: command not found
PS C:\Users\atnnn> wsl --shell-type none /bin/sh -c '/nix/var/nix/profiles/system/sw/bin/ln -s /nix/var/nix/profiles/system/sw/bin/getent /bin/getent'
PS C:\Users\atnnn> wsl
[nixos@nixos:/mnt/c/Users/atnnn]$
```

**Comment by nzbr** [comment]
It seems that `getent` was moved out of the `glibc.bin` derivation on unstable. It should be fixed by #267.

As for `wslpath` missing from PATH: It will be fixed with the upcoming syschdemd replacement, so it doesn't really make sense to fix it for the soon to be replaced version
